### PR TITLE
Lookup table by id

### LINKED ISSUE #625: Introduce reference maps to do lookup
Currently, we lookup for definitions via `lookupValue`:

https://github.com/informalsystems/quint/blob/7c29089ab746c4d925eaf423a2e96b4619e14245/quint/src/lookupTable.ts#L120-L138

Since names do not have to be unique within a module or an operator definition, `lookupValue` does an additional search that relates the referencing expression to the definition. This may be a potential performance bottleneck (we don't know whether it is an actual bottleneck). Instead of doing the lookup via names and ids, we could simply store a map of identifiers to identifiers, which maps a reference id to the id of the definition that has introduced the name. This would make the lookup conceptually easier and, probably, more efficient.

This work is blocked by #624.

### PR DESCRIPTION
**Pull request by bugarela**
Hello :octocat: 

This closes #625 by replacing our current lookup strategy that took name and scope with a new one that takes only the expression ID, in all phases after parse phase 2 (name resolution).

We still use the old table, now called `DefinitionsByName` to resolve imports, and the new lookup table is constructed during name resolution.

Most of the changes in this PR are adaptations for existing code to always do lookup by id instead of name, since the ID was not available in some places. For the REPL, we do still index some information by name in order to retrieve it by name (such as `_lastTrace`). The names for which we do this are defined in the `specialNames` list.

- [ ] Tests added for any new code
- [ ] Ran `npm run format` (or had formatting run automatically on all files edited)
- [ ] Documentation added for any new functionality
- [ ] Entries added to the respective `CHANGELOG.md` for any new functionality
- [ ] Feature table on [`README.md`](../README.md#roadmap) updated for any listed functionality

### CODE REVIEW

File: quint/src/definitionsByName.ts
Comment (bugarela) [inline quint/src/definitionsByName.ts:1]: This is all moved from `lookupTable.ts`. Behavior is unchanged, I just renamed some things and updated the comments.

File: quint/src/effects/inferrer.ts [hunk 1]
```diff
@@ -63,73 +59,46 @@ export class EffectInferrer implements IRVisitor {
...
+  /*  { kind: 'var', identifier: v } ∈ Γ
+   * ------------------------------------ (NAME-VAR)
+   *          Γ ⊢ v: Read[v]
+   */
+  exitVar(def: QuintVar) {
+    const effect: Effect = {
+      kind: 'concrete', components: [{ kind: 'read', entity: { kind: 'concrete', stateVariables: [{ name: def.name, reference: def.id }] } }],
+    }
+    this.addToResults(def.id, right(toScheme(effect)))
+  }
```
Comment (bugarela) [inline quint/src/effects/inferrer.ts:90]: I'm now adding the type/effects to the const/var definitions themselves and then name expressions only propagate these type/effects by using the lookup table. The implementation for `exitName` is significantly simplified for both type and effect inference.

File: quint/src/runtime/impl/compilerImpl.ts [hunk 2]
```diff
@@ -1231,18 +1250,37 @@ export class CompilerVisitor implements IRVisitor {
...
+    const nruns = mkConstComputable(rv.mkInt(1n))
+    this.runTestSimulation(nruns, nsteps, init, next, inv)
+  }
+
+  // The simulator core: produce multiple random runs
+  // and check the given state invariant (state assertion).
+  //
+  // Technically, this is similar to the implementation of folds.
+  // However, it also restores the state and saves a trace, if there is any.
+  private runTestSimulation(
```
Comment (bugarela) [inline quint/src/runtime/impl/compilerImpl.ts:1281]: Extracted from `private test` in order to reuse it in `private testOnce`. Mostly left unchanged except that now we don't lookup computables for `init`, `next` and `inv` anymore, as they are computables themselves.

**Review by konnov — APPROVED** [review]
This is an amazing improvement. I believe that it will simplify our maintenance costs in the future

File: quint/src/definitionsByName.ts [hunk 3]
```diff
@@ -0,0 +1,216 @@
+import { QuintType } from "./quintTypes"
```
[the remark this answers is not in the document]
Comment (bugarela) [reply quint/src/definitionsByName.ts:1]: Yes! I thought the same thing when creating it, but renaming changed files can get git diffs pretty crazy, so I postponed this. I think we should have `nameResolver.ts`, `definitionsCollector.ts`, `definitionsScanner.ts` and `lookupTable.ts` in this directory as well.

I'll do that in an isolated PR after this is merged.

File: quint/src/lookupTable.ts [hunk 4]
```diff
@@ -1,215 +1,24 @@
...
 }
 
 /**
- * Merges two lookup tables together in a new one.
+ * A lookup table from IR component ids to their definitions.
+ * Should have an entry for every IR component with a name
+ * That is:
+ * - Name expressions
+ * - App expressions (opcode is a name)
+ * - Override parameters in instance definitions
```
[the remark this answers is not in the document]
Comment (bugarela) [reply quint/src/lookupTable.ts:19]: Actually, this is not right. Parameters themselves are not in the lookup table as they would point to themselves, which is useless. The entries in the lookup table are actually the name expressions or applications that refer to those parameters (i.e. in `def a(f, x) = f(x)`, the `f` being applied and its `x` parameter will point to the parameters of `a`, but the parameters of `a` won't point to anything.

[the remark this answers is not in the document]
Comment (konnov) [reply quint/src/lookupTable.ts:19]: oh, that's true. I have not thought about that :)

**Comment by konnov** [comment]
Let's merge this PR and see it running!

**Pull request merged**